File: backend/src/main/java/com/societyos/society/entity/Complaint.java
package com.societyos.society.entity;

import com.societyos.user.entity.User;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(
        name = "society_complaints",
        indexes = {
                @Index(
                        name = "idx_society_complaints_society",
                        columnList = "society_id"
                ),
                @Index(
                        name = "idx_society_complaints_creator",
                        columnList = "created_by"
                ),
                @Index(
                        name = "idx_society_complaints_status",
                        columnList = "status"
                ),
                @Index(
                        name = "idx_society_complaints_assigned_to",
                        columnList = "assigned_to"
                )
        }
)
@Getter
@Setter
@NoArgsConstructor
public class Complaint {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "society_id", nullable = false)
    private Society society;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "created_by", nullable = false)
    private User createdBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "unit_id")
    private Unit unit;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private ComplaintCategory category;

    @Column(nullable = false, length = 150)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ComplaintPriority priority;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private ComplaintStatus status;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_to")
    private User assignedTo;

    @Column(name = "resolution_note", columnDefinition = "TEXT")
    private String resolutionNote;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @Column(name = "resolved_at")
    private OffsetDateTime resolvedAt;

    @PrePersist
    protected void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        createdAt = now;
        updatedAt = now;

        if (status == null) {
            status = ComplaintStatus.OPEN;
        }

        if (priority == null) {
            priority = ComplaintPriority.MEDIUM;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}

File: frontend/src/types/complaint.ts
export type ComplaintCategory =
  | 'MAINTENANCE'
  | 'SECURITY'
  | 'CLEANLINESS'
  | 'WATER'
  | 'ELECTRICITY'
  | 'PARKING'
  | 'NOISE'
  | 'OTHER'

export type ComplaintPriority =
  | 'LOW'
  | 'MEDIUM'
  | 'HIGH'
  | 'URGENT'

export type ComplaintStatus =
  | 'OPEN'
  | 'IN_PROGRESS'
  | 'RESOLVED'
  | 'CLOSED'
  | 'CANCELLED'

export interface CreateComplaintRequest {
  category: ComplaintCategory
  title: string
  description: string
  priority?: ComplaintPriority
  unitId?: string
}

export interface ComplaintResponse {
  id: string
  societyId: string
  createdBy: string
  createdByEmail: string
  unitId?: string
  unitNumber?: string
  category: ComplaintCategory
  title: string
  description: string
  priority?: ComplaintPriority
  status: ComplaintStatus
  assignedTo?: string
  assignedToEmail?: string
  resolutionNote?: string
  createdAt: string
  updatedAt: string
  resolvedAt?: string
}